### ISSUE TITLE
Beam sync stalls

### ISSUE THREAD

**Issue by cburgdorf** [ethereum/trinity#1806]
### What is wrong?

After a while beam sync stops importing blocks. In the attached log line `80319` is the last time it successfully imported a block. At line `86426` it is the first time that the `BaseBodyChainSyncer` reports it has no write capacity.

```
DEBUG  2020-06-20 09:45:52,898   BaseBodyChainSyncer  (progress, queued, max) of bodies, receipts: [(0, 0, 2048)]. Write capacity? False Importing? True
```
Another interesting thing going on is that there are quite a lot of timed out requests for state data and they start shortly after the last successful block import on line `80663`

```
Timed out requsting state data for block #2906107, retrying...
```

[trinity_dappnode.zip](https://github.com/ethereum/trinity/files/4809079/trinity_dappnode.zip)

Notice this is running:

1. On Görli
2. Against a local Geth Görli node
3. On #1805 which fixes a crash that would otherwise take the whole node down reliably

### How can it be fixed

Need to dig in.

**Comment by carver** [comment]
So the problem seems to start here:
```
   DEBUG  2020-06-20 09:37:39,000        BeamDownloader  ETHPeer (eth, 65) <Session <Node(0x6f621a@91.13.207.112)> cbfe5ab6-fcc7-4b73-9e5c-cc20b4f1f018> returned no urgent nodes from (HexBytes('0x423c7cec6a4b0bec1ea9d78e5f0d4551f9ddb8cc23e87aa8c83a9a4cdf4a38da'),)
```

^ Against a known local node, when beam sync isn't lagging, I wouldn't ever expect to get that message.

It means that beam sync is missing some critical trie node and the peer didn't supply the node body as an answer to a `GetNodeData`.

So that's its own problem that needs to be fixed. But another problem is: it should have been retrying. I don't see any evidence in the logs that `BeamDownloader` made any more state requests. I expect it to keep attempting (because maybe it can get the node from a different peer, or maybe the peer is also beam syncing and will have the node available later). But there were no more `GetNodeData` requests after the "returned no urgent nodes" log line, that I can see.

I scanned the code, and didn't immediately see why that was happening. So I would say this should be split into two issues:
1. For some reason, on Goerli block `<BlockHeader #2906099 f3245656>` there is a missing node that peers don't recognize as available. It appears to happen during persist (because the block reward check succeeds, and all that's left is to persist the state. It's normal to sometimes be missing state during persist, due to trie refactors. Hunting down the solution can be quite tricky though. -- I'm a bit afraid that we might have to make major overhauls to py-trie in how batches of changes are applied to avoid this problem, but won't jump to that conclusion yet
2. When an urgent node is missing, we should mark it for re-attempt. -- I have a lead. I think `QueeningQueue` might have a bug where if your queen gets penalized and then you never get another peer besides the queen coming back, then you can hang on an `await queening_queue.get_queen_peer()` call, even though the queen came back.

**Comment by carver** [comment]
Pushed a fix for the second issue, but unfortunately it doesn't deal with the first issue at all.

**Comment by carver** [comment]
Cool, the good news is that I have a fully synced trinity node (so, archive). It doesn't have a trie node for hash `HexBytes('0x423c7cec6a4b0bec1ea9d78e5f0d4551f9ddb8cc23e87aa8c83a9a4cdf4a38da')`, so I should have an easy-to-reproduce case to hunt down the bug, now. The full DB is 43GB, and it took about the whole last 5 days to sync from scratch (plus some custom local patches). If anyone else wants to take a look, too, I can transfer a zip of the DB to you, to save you the sync time.

**Comment by cburgdorf** [comment]
That's interesting. I'm really curious about this bug! Having Görli as a thin, light testnet is really helpful for these kind of issues.

**Comment by carver** [comment]
The bad news is that if I force beam sync to start at block `#2906090` (picked arbitrarily before the problem block), then `#2906099` (the problem block) succeeds and I don't reproduce the bug. :(

The log linked in the main issue body says that it started beam importing at `#2905984`, so I'm trying that next... *and hooray, that reproduces the problem!*

**Comment by carver** [comment]
Ew, this is going to be an ugly one. After it fails out at `#2906099`, if I restart beam sync from that block, then the import succeeds. So far, the only way I know how to reproduce this is to start back at `#2905984`. (... and I'm not even sure that will work reliably)

**Comment by carver** [comment]
Wow, it gets weirder:

>    ERROR  2020-06-29 16:30:14,788        BeamDownloader  Never collected node data for hashes {HexBytes('0x423c7cec6a4b0bec1ea9d78e5f0d4551f9ddb8cc23e87aa8c83a9a4cdf4a38da')}

Which normally means that Beam Sync is getting stuck... but it is continuing to import.

**Comment by carver** [comment]
So based on the stack trace: `persist()` is raising a (typical) `MissingStorageTrieNode` while building the state root. The fact that it is only sometimes reproducible suggests to me that maybe it is actually raising this exception multiple times, and retrying. The fact that we're getting an error at all suggests that maybe after the first exception the trie is left in some kind of funny state that it does not recover from cleanly, and something about being in that funny state means that a locally-built trie node (which is only generated in the middle of building the trie for the final state root) is getting dropped between re-runs.

Now, how to confirm and test this? Well, since I don't know the precise parameters that trigger the scenario, and generating the failure is hit-or-miss, it's going to be tricky. A couple options:
- try to guess at the kind of scenario that generates and deletes an intermediate trie node (this could take a while)
- keep re-running, while adding logs, and trying enough info about the failure (this could also take a while)

**Comment by carver** [comment]
Oof, this one is proving very difficult to replicate. I think I got lucky the first time. Or worse: adding debug logs is actually making it harder to reproduce. :(

I've just run it ~10 times in a row with no reproduction... Now that I have more helpful debug logs in place.

**Comment by carver** [comment]
Well the resolution is simultaneously aggravating and relieving. The bug is on the py-trie side, and was fixed in py-trie v2.0.0a2. It reliably reproduces in alpha 1, and I cannot reproduce in alpha 2. There are a couple fixes in alpha 2, but since this is a bug where a locally created node is deleted even when it's still needed, it seems to probably be this one:
https://github.com/ethereum/py-trie/blob/e3889c54013125d0f664fbfa6a4dea4bc85e67a6/CHANGELOG#L35-L38

Which was merged in https://github.com/ethereum/trinity/pull/1818

There's just enough complication in reproducing it that I don't think it's worth tracking down and building its own test case on the py-evm/trinity side. The bug was completely in the py-trie domain.
